File: problems/0014-longest-common-prefix/solution.go
package leetcode

import "slices"

func longestCommonPrefix(strs []string) string {
	first := strs[0]
	for i := range len(first) {
		for _, s := range strs[1:] {
			if i == len(s) || s[i] != first[i] {
				return first[:i]
			}
		}
	}
	return first
}

func longestCommonPrefixMinMax(strs []string) string {
	lo, hi := slices.Min(strs), slices.Max(strs)
	for i := range len(lo) {
		if lo[i] != hi[i] {
			return lo[:i]
		}
	}
	return lo
}

File: problems/0014-longest-common-prefix/solution.ts
function longestCommonPrefix(strs: string[]): string {
  const [first = ""] = strs;
  for (let i = 0; i < first.length; i++) {
    if (!strs.every((s) => s[i] === first[i])) {
      return first.slice(0, i);
    }
  }
  return first;
}
